### PR TITLE
Bug/gcra race

### PR DESCRIPTION
**Pull request by Jude188**
Fixes Issue #5 

TODO:
  * Add tests to restore test coverage [DONE]
  * Implement `RedisStore.compare_and_swap` [DONE]

### CODE REVIEW

File: src/rush/limiters/gcra.py [hunk 1]
```diff
@@ -24,62 +30,82 @@ def rate_limit(
...
-                used=used, remaining=remaining, time=new_time
-            )
-        else:
-            limitdata = limit_data.LimitData(
-                used=used, remaining=remaining, created_at=now, time=new_time
+        attempts = 0
+        while attempts < MAX_CAS_ATTEMPTS:
+            # tat is short for theoretical arrival time, we store this as
+            # "time" on our limit data
+            tat = getattr(data, "time", None) or now
```
Comment (sigmavirus24) [inline src/rush/limiters/gcra.py]: I'd prefer to make this a separate method

Comment (Jude188) [reply src/rush/limiters/gcra.py]: Do you mean have a function to check whether to retry which takes the number of attempts and returns a bool? 

Or that you'd prefer to have the contents of the while loop split out into a separate function?

Comment (sigmavirus24) [reply src/rush/limiters/gcra.py]: The latter

Comment (Jude188) [reply src/rush/limiters/gcra.py]: Done. As noted in the commit message, the way I've done this does mean that we recalculate `emission_interval` and `delay_variation_tolerance` an extra time for each CAS failure. I suspect this isn't the end of the world though as it's only only division and one multiplication and it saves us passing two additional arguments to `_gcra` which seems much neater.

Comment (Jude188) [reply src/rush/limiters/gcra.py]: Apologies for squashing the commit just as you reviewed! Makes it a bit tricky to see the further changes I made to this bit of the code.

**Comment by Jude188** [comment]
@sigmavirus24

**Comment by Jude188** [comment]
@sigmavirus24 bump

**Pull request closed without merging**